internal/querynodev2/segments/metricsutil/observer.go

```diff
@@ -0,0 +1,245 @@
+			ticker := time.NewTicker(d)
+			defer ticker.Stop()
+			for range ticker.C {
+				expireAt := time.Now().Add(-d)
+				globalObserver.Expire(expireAt)
+			}
+		}()
+	})
+	return globalObserver
+}
+
+// newSegmentsObserver creates a new segmentsObserver.
+// Used to check if a segment is hot or cold.
+func newSegmentsObserver() *segmentsObserver {
+	return &segmentsObserver{
+		nodeID:   strconv.FormatInt(paramtable.GetNodeID(), 10),
+		segments: typeutil.NewConcurrentMap[SegmentLabel, *segmentObserver](),
+	}
+}
+
+// segmentsObserver is a observer all segments metrics.
+type segmentsObserver struct {
+	nodeID   string
+	segments *typeutil.ConcurrentMap[SegmentLabel, *segmentObserver] // map segment id to observer.
+	// one segment can be removed from one query node, for balancing or compacting.
+	// no more search operation will be performed on the segment after it is removed.
+	// all related metric should be expired after a while.
+	// may be a huge map with 100000+ entries.
+}
+
+// Observe records a new metric
+func (o *segmentsObserver) Observe(m labeledRecord) {
+	if m.getError() != nil {
+		return // reject error record.
+		// TODO: add error as a label of metrics.
+	}
+	// fast path.
+	label := m.Label()
+	observer, ok := o.segments.Get(label)
+	if !ok {
+		// slow path.
+		newObserver := newSegmentObserver(o.nodeID, label)
+		observer, _ = o.segments.GetOrInsert(label, newObserver)
+	}
+	// do a observer.
+	observer.Observe(m)
+}
+
+// Expire expires the observer.
+func (o *segmentsObserver) Expire(expiredAt time.Time) {
+	o.segments.Range(func(label SegmentLabel, value *segmentObserver) bool {
+		if value.IsExpired(expiredAt) {
+			o.segments.Remove(label)
+			value.Clear()
+			return true
+		}
+		return true
+	})
+}
+
+// newSegmentObserver creates a new segmentObserver.
+func newSegmentObserver(nodeID string, label SegmentLabel) *segmentObserver {
+	now := time.Now()
+	return &segmentObserver{
+		label:       label,
+		prom:        newPromObserver(nodeID, label),
+		lastUpdates: atomic.NewPointer[time.Time](&now),
+	}
+}
+
+// segmentObserver is a observer for segment metrics.
+type segmentObserver struct {
+	label SegmentLabel // never updates
+	// observers.
+	prom promMetricsObserver // prometheus metrics observer.
+	// for expiration.
+	lastUpdates *atomic.Pointer[time.Time] // update every access.
+}
+
+// IsExpired checks if the segment observer is expired.
+func (o *segmentObserver) IsExpired(expireAt time.Time) bool {
+	return o.lastUpdates.Load().Before(expireAt)
+}
+
+// Observe observe a new
+func (o *segmentObserver) Observe(m labeledRecord) {
+	now := time.Now()
+	o.lastUpdates.Store(&now)
+
+	switch mm := m.(type) {
+	case *CacheLoadRecord:
+		o.prom.ObserveCacheLoad(mm)
+	case *CacheEvictRecord:
+		o.prom.ObserveCacheEvict(mm)
+	case QuerySegmentAccessRecord:
+		o.prom.ObserveQueryAccess(mm)
+	case SearchSegmentAccessRecord:
+		o.prom.ObserveSearchAccess(mm)
+	default:
+		panic("unknown segment access metric")
```

Comment:
just log.Warn?

Reply:
it's just an assertion.
it's an unreachable code branch.
we can find bug when develop and CI by panic message easily.